C1 - Mailroom OO Revision

natasha-aleksandrova opened the pull request:
Please review the revision. Anything else should be added? (or removed?)

PythonCHB commented on source/exercises/mailroom-oo.rst:
```diff
@@ -51,29 +52,64 @@ One of the hardest parts of OO design (particularly in Python) is to know how "l
+Below are more detailed suggestions on breaking down your existing code into multiple modules that will be part of a single mailroom program.
+
+Donor Class
+^^^^^^^^^^^^
```
RST allows pretty much any charactor to make headings -- but I've tried ot be consistent in these docs with:

`======`
then
`--------`
then 
`........`

so I think this level should be:
`..............`

PythonCHB replied:
BTW -- are you building the docs?

running `make html` in the root dir of the repo should build them for you locally.

natasha-aleksandrova replied:
ah good to know! will update heading.

no I did not run `make html`... I feel like you've mentioned this before and ended up doing it for me last time. I'll see if I can figure it out. (haven't worked with this type of docs before).

FYI I'm using Pycharm with a plugin to render these.

PythonCHB commented on source/exercises/mailroom-oo.rst:
```diff
@@ -51,29 +52,64 @@ One of the hardest parts of OO design (particularly in Python) is to know how "l
+DonorCollection Class
+^^^^^^^^^^^^^^^^^^^^^
+
+Module responsible for donor collection data encapsulation.
```
we should be clear about what is in a "module" vs a "class".

In python, you can have one class per module, but that's not necessarily a common practice.

But in any case, we should be clear in the description.

natasha-aleksandrova replied:
Do you think this should be renamed to Class instead of module? Or explanation at the top about module vs class?

PythonCHB commented on source/exercises/mailroom-oo.rst:
```diff
@@ -51,29 +52,64 @@ One of the hardest parts of OO design (particularly in Python) is to know how "l
+* ``print`` statements to print to console.
+
+
+.. note::  Technically console print statement don't belong in your data classes, however for some features like send letters instead of "sending" we are simply printing so it is ok for this feature to reside in the data class.
```
I try to have them have all the methods in the data classes return strings, rather than call print() directly.

But yes, at _some_point, you need to print -- but that could be in the CLI or Data classes.

PythonCHB commented:
In my OO mailroom solution:
`source/solutions/mailroom/mailroom_oo`

I kept everything in one file (plus a test file).

Then, in mailroom_package (in course 2, though it may not be there anymore), I split it up similarily to how you are suggesting:

`source/solutions/mailroom/mailroom_pkg`

While also adding a bit more to make it a proper package.

I like the idea of having them split up the data management and CLI, but I wonder if we'll have confusion about how to import, and how to run the code.

It may be just fine if they always run it from the dir where the files are -- but be prepared for questions!

PythonCHB commented on source/exercises/mailroom-oo.rst:
```diff
@@ -51,29 +52,64 @@ One of the hardest parts of OO design (particularly in Python) is to know how "l
 
-3) If a function works with multiple donors -- it should be in the class that handles a collection of donors.
+For example, you may create 3 test files:
+``test_donor.py``, ``test_donor_collection.py`` and ``test_cli_main.py`` - each testing corresponding module.
```
we may not want them to tryt o test the cli -- there isn't much to test without mocking input() -- and that's too much to ask at this stage (though I have had students figure it out).

I got into that in course two -- advanced testing.

natasha-aleksandrova replied:
Ok I can remove test_cli_main reference as it should be optional at this point since it would be heavy on  mocks and although some students do use mocks in my py210, some are very confused about them 🙂

natasha-aleksandrova commented:
for your main comment; given that they have to import mailroom to test it in lesson 6 (along with all further assignments that unit test), they are doing imports already.

natasha-aleksandrova commented:
@PythonCHB thanks for review!

I will work on updates later today.

PythonCHB commented:
Yes -- for importing, but we should probably put a bit of guidance in here, too to remind them.

As for building the docs -- it's using Sphinx, which is pretty simple to use (once it's set up...), but you do need to have Sphinx and rtd_theme installed. I think this will do it:

`pip install sphinx sphinx_rtd_theme`

I suspect PyCharm is doing raw RST rendering, which gets you most of the way, but won't have any of the Sphinx niceties like table of contents, cross referencing, etc.

natasha-aleksandrova commented:
@PythonCHB ok I have pushed some updates.

* fixed the heading
* simplified mailroom code organization since that is something they do in next course anyway

FYI, I tried installing Sphinx and running `make html` but running into missing dependency errors. I'll see if I can get away with pycharm plugin so I don't get bogged down on this.

PythonCHB commented:
I added a touch -- but it looks great -- merging away.

natasha-aleksandrova commented:
great adds! thanks!

natasha-aleksandrova commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Mailroom OO](https://trello.com/c/cg3lsV1k/1-mailroom-oo)
